metastore/src/java/org/apache/hadoop/hive/metastore/ObjectStore.java

```diff
@@ -8268,29 +8279,90 @@ public NotificationEventResponse getNextNotification(NotificationEventRequest rq
     }
   }
 
+  private void lockForUpdate() throws MetaException {
+    String selectQuery = "select NEXT_EVENT_ID from NOTIFICATION_SEQUENCE";
+    String selectForUpdateQuery = sqlGenerator.addForUpdateClause(selectQuery);
+    new RetryingExecutor(hiveConf, () -> {
+      Query query = pm.newQuery("javax.jdo.query.SQL", selectForUpdateQuery);
+      query.setUnique(true);
+      // only need to execute it to get db Lock
+      query.execute();
+    }).run();
+  }
+
+  static class RetryingExecutor {
+    interface Command {
+      void process() throws Exception;
+    }
+
+    private static Logger LOG = LoggerFactory.getLogger(RetryingExecutor.class);
+    private final int maxRetries;
+    private final long sleepInterval;
+    private int currentRetries = 0;
+    private final Command command;
+
+    RetryingExecutor(Configuration config, Command command) {
+      this.maxRetries = config.getInt(ConfVars.NOTIFICATION_SEQUENCE_LOCK_MAX_RETRIES.name(),
+          ConfVars.NOTIFICATION_SEQUENCE_LOCK_MAX_RETRIES.defaultIntVal);
+      this.sleepInterval = config.getTimeDuration(
+          ConfVars.NOTIFICATION_SEQUENCE_LOCK_RETRY_SLEEP_INTERVAL.name(),
+          ConfVars.NOTIFICATION_SEQUENCE_LOCK_RETRY_SLEEP_INTERVAL.defaultLongVal,
+          TimeUnit.MILLISECONDS
+      );
+      this.command = command;
+    }
+
+    public void run() throws MetaException {
+      while (true) {
+        try {
+          command.process();
+          break;
+        } catch (Exception e) {
+          LOG.info("retried / max : {} / {}", currentRetries, maxRetries);
+          if (currentRetries >= maxRetries) {
+            LOG.error("failed to execute Command :: ", e);
+            throw new MetaException(e.getMessage());
+          }
+          currentRetries++;
+          try {
+            Thread.sleep(sleepInterval);
+          } catch (InterruptedException e1) {
+            String msg = "error waiting for next retry " + currentRetries;
+            LOG.error(msg, e1);
+            throw new MetaException(msg);
+          }
+        }
+      }
+    }
+  }
+
   @Override
   public void addNotificationEvent(NotificationEvent entry) {
     boolean commited = false;
     Query query = null;
     try {
       openTransaction();
-      query = pm.newQuery(MNotificationNextId.class);
-      Collection<MNotificationNextId> ids = (Collection) query.execute();
-      MNotificationNextId id = null;
+      lockForUpdate();
+      Query objectQuery = pm.newQuery(MNotificationNextId.class);
```

Comment:
What if we use the same query from lockForUpdate() here instead of making two calls? The lockForUpdate is getting the next ID as well as this query.

Reply:
wanted to do the above as we are using the retried object to persist the new value as well, which if done via the lockForUpdate would have to be done explicitly on the same NNI_ID row value